### PR TITLE
planner, executor: parallelize stream agg by utilizing shuffle executor

### LINKED ISSUE #20651: Parallelize stream aggregation executors by using shuffle executors
## Description

In this PR(https://github.com/pingcap/tidb/pull/14238), we introduce the shuffle executor.
Like https://github.com/pingcap/tidb/issues/14441, we may use this shuffle executor to implement parallel stream aggregation by nesting some sort and stream-aggg executors into a shuffle executor:
```
ShuffleExec
    - StreamAgg
        - Sort
    - StreamAgg
        - Sort
    ...
``` 
It can utilize more CPU resources to get better performance.

- [x] Introduce a new variable to control concurrency
- [x] Support `shuffle + stream_agg` and add some tests and benchmarks simply
- [ ] Add more tests and benchmarks about `shuffle + stream_agg`
- [x] Support `range_spliter`

## Score

6600

## Mentor

@qw4990 (Slack ID: Zhang Yuanjia)

## Recommend Skills

- Golang

### PR DESCRIPTION
**Pull request by 3AceShowHand**
### What problem does this PR solve?

Issue Number: close #20651

Problem Summary: This pr make stream aggregation run in a parallel way by using shuffle executor

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

* planner: build the physical plan for stream aggregation with shuffle plan.
* executor: build stream aggregation executor with shuffle executor.

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- No release note

### CODE REVIEW

File: executor/aggregate.go [hunk 1]
```diff
@@ -873,6 +873,9 @@ func (e *StreamAggExec) Close() error {
 // Next implements the Executor Next interface.
 func (e *StreamAggExec) Next(ctx context.Context, req *chunk.Chunk) (err error) {
 	req.Reset()
+	// if e.executed {
+	// 	logutil.BgLogger().Info(fmt.Sprintf("StreamAggExec end %v", e.partitionId))
+	// }
```
Comment (qw4990) [inline executor/aggregate.go:878]: Please remove these lines.

File: config/config.toml.example [hunk 2]
```diff
@@ -32,8 +32,8 @@ split-table = true
 token-limit = 1000
 
 # The maximum memory available for a single SQL statement. Default: 1GB
+# mem-quota-query = 53687091200
 mem-quota-query = 1073741824
-
```
Comment (qw4990) [inline config/config.toml.example:36]: Please remove these changes.

File: executor/analyze_test.go [hunk 3]
```diff
@@ -123,7 +123,7 @@ func (s *testSuite1) TestAnalyzeReplicaReadFollower(c *C) {
 	tk.MustExec("analyze table t")
 }
 
-func (s *testSuite1) TestClusterIndexAnalyze(c *C) {
+func (s *testSerialSuite1) TestClusterIndexAnalyze(c *C) {
```
Comment (qw4990) [inline executor/analyze_test.go:126]: We have decided to skip this test now https://github.com/pingcap/tidb/pull/20815.

File: sessionctx/variable/tidb_vars.go [hunk 4]
```diff
@@ -524,9 +531,11 @@ const (
 	DefTiDBHashAggPartialConcurrency   = ConcurrencyUnset
 	DefTiDBHashAggFinalConcurrency     = ConcurrencyUnset
 	DefTiDBWindowConcurrency           = ConcurrencyUnset
+	DefTiDBStreamAggConcurrency        = ConcurrencyUnset
```
Comment (qw4990) [inline sessionctx/variable/tidb_vars.go:534]: How about disabling this feature by default?

File: util/chunk/row_container.go [hunk 5]
```diff
@@ -80,6 +80,7 @@ func (c *RowContainer) SpillToDisk() {
 	N := c.m.records.NumChunks()
 	c.m.recordsInDisk = NewListInDisk(c.m.records.FieldTypes())
 	c.m.recordsInDisk.diskTracker.AttachTo(c.diskTracker)
+	// logutil.BgLogger().Info(fmt.Sprint("SpillToDisk", N, c.m.records.NumRowsOfChunk(0)))
```
Comment (qw4990) [inline util/chunk/row_container.go:83]: Please remove this line.

File: planner/core/plan.go [hunk 6]
```diff
@@ -92,10 +92,44 @@ func optimizeByShuffle(pp PhysicalPlan, tsk task, ctx sessionctx.Context) task {
...
+		Tail:         tail,
+		DataSource:   dataSource,
+		SplitterType: PartitionHashSplitterType,
+		ByItems:      pp.GroupByItems,
+	}.Init(ctx, pp.statsInfo(), pp.SelectBlockOffset(), reqProp)
+	return shuffle
+}
+
+func getPartitionSplitter(ctx sessionctx.Context) PartitionSplitterType {
+	if ctx.GetSessionVars().IsStreamAggRangePartitionEnabled() {
```
Comment (qw4990) [inline planner/core/plan.go:127]: I think we can remove this variable and always use `RangeSplitter` after it is implemented.

File: sessionctx/variable/tidb_vars.go [hunk 7]
```diff
@@ -524,9 +531,11 @@ const (
 	DefTiDBHashAggPartialConcurrency   = ConcurrencyUnset
 	DefTiDBHashAggFinalConcurrency     = ConcurrencyUnset
 	DefTiDBWindowConcurrency           = ConcurrencyUnset
+	DefTiDBStreamAggConcurrency        = 1
 	DefTiDBForcePriority               = mysql.NoPriority
 	DefTiDBUseRadixJoin                = false
 	DefEnableWindowFunction            = true
+	DefEnableStreamAggRangePartition   = false
```
Comment (qw4990) [inline sessionctx/variable/tidb_vars.go:538]: We can remove this flag and always use `RangePartition` after it is implemented.

File: planner/core/testdata/plan_suite_out.json [hunk 8]
```diff
@@ -56,7 +56,8 @@
       },
       {
         "SQL": "select /*+ STREAM_AGG() */ s, count(s) from (select /*+ HASH_AGG() */ sum(t1.a) as s from t t1, t t2 where t1.a = t2.b group by t1.a) p group by s",
-        "Best": "LeftHashJoin{IndexReader(Index(t.f)[[NULL,+inf]])->TableReader(Table(t))}(test.t.a,test.t.b)->Projection->HashAgg->Sort->StreamAgg->Projection"
+        "Best":"LeftHashJoin{IndexReader(Index(t.f)[[NULL,+inf]])->TableReader(Table(t))}(test.t.a,test.t.b)->Projection->HashAgg->Sort->StreamAgg->Projection"
+
```
Comment (qw4990) [inline planner/core/testdata/plan_suite_out.json:60]: Please remove this change.

File: planner/core/plan.go [hunk 9]
```diff
@@ -92,10 +92,44 @@ func optimizeByShuffle(pp PhysicalPlan, tsk task, ctx sessionctx.Context) task {
...
+func optimizeByShuffle4StreamAgg(pp *PhysicalStreamAgg, ctx sessionctx.Context) *PhysicalShuffle {
+	concurrency := ctx.GetSessionVars().StreamAggConcurrency()
+	if concurrency <= 1 {
+		return nil
+	}
+
+	sort, ok := pp.Children()[0].(*PhysicalSort)
+	if !ok {
+		return nil
+	}
```
Comment (qw4990) [inline planner/core/plan.go:112]: This check can be removed.

File: executor/aggregate_test.go [hunk 10]
```diff
@@ -1142,3 +1145,127 @@ func (s *testSuiteAgg) TestIssue17216(c *C) {
...
+		}
+	}
+}
+
+func (s *testSuiteAgg) TestIssue20658(c *C) {
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec("use test;")
+
+	aggFuncs := []string{"count(a)", "sum(a)", "avg(a)", "max(a)", "min(a)", "bit_or(a)", "bit_xor(a)", "bit_and(a)"}
+	aggFuncs2 := []string{"var_pop(a)", "var_samp(a)", "stddev_pop(a)", "stddev_samp(a)", "approx_count_distinct(a)", "approx_percentile(a, 7)"}
```
Comment (qw4990) [inline executor/aggregate_test.go:1210]: Why not combine them into one?

File: executor/benchmark_test.go [hunk 11]
```diff
@@ -381,6 +433,21 @@ func BenchmarkAggRows(b *testing.B) {
 	}
 }
 
+func BenchmarkHashAggRows(b *testing.B) {
```
Comment (qw4990) [inline executor/benchmark_test.go:436]: Should be `BenchmarkShuffleStreamAgg`?

File: executor/aggregate_test.go (hunk 10 above)
Comment (JinLingChristopher) [reply executor/aggregate_test.go:1210]: for group_concat, it's hard to get reconstruct the result.

File: executor/shuffle.go [hunk 12]
```diff
@@ -71,7 +71,7 @@ import (
 //          +---------->  |    fetch data from DataSource   |
 //                        +---------------------------------+
 //
-////////////////////////////////////////////////////////////////////////////////////////
```
Comment (qw4990) [inline executor/shuffle.go:74]: Please remove all unnecessary changes in this file.
